File: csharp/spiral-matrix/SpiralMatrix.cs
using System;
using System.Numerics;

public class SpiralMatrix
{
    // inspired by JakDrako on https://www.reddit.com/r/dailyprogrammer/comments/6i60lr/20170619_challenge_320_easy_spiral_ascension/
    public static int[,] GetMatrix(int size)
    {
        int[,] matrix = new int[size, size];
        var position = new Complex(-1, 0); // start outside of the "board"
        var direction = new Complex(1, 0); // heading to the right
        int numSpiralPartsWithThisSize = 1; // e.g. for n=3 the lengths of straight spiral parts are 3, 2, 2, 1, 1
        int counter = 1;
        while (size > 0)
        {
            for (int i = 0; i < size; i++) // walk one such straight part
            {
                position += direction;
                matrix[(int)position.Imaginary, (int)position.Real] = counter; // in order to get it right, the ROWS of the matrix (first index)
                // have to vary with the imaginary part, while the COLUMS (second index) correspond to the real part
                counter++;
            }
            direction *= Complex.ImaginaryOne; // now rotate 90 degrees to the right
            if (--numSpiralPartsWithThisSize == 0) // look if there are straight parts left
            {
                numSpiralPartsWithThisSize = 2;
                size--;
            }
        }
        return matrix;
    }
}

File: fsharp/spiral-matrix/SpiralMatrix.fs
module SpiralMatrix

open System.Numerics

let spiralMatrix (size: int) = 
    let matrix = Array2D.zeroCreate size size
    // let len = matrix.GetLength 0
    let mutable leng = size
    let mutable position = Complex(0., -1.)
    let mutable direction = Complex(0., 1.)
    let mutable numSpiralPartsWithSize = 1
    let mutable counter = 1
    while leng > 0 do
        for i in [0..(leng - 1)] do
            position <- (position + direction)
            matrix.[int (position.Real), int (position.Imaginary)] <- counter
            counter <- (counter + 1)
        direction <- (direction * (- Complex.ImaginaryOne))
        if numSpiralPartsWithSize = 1 then
            numSpiralPartsWithSize <- 2
            leng <- (leng - 1)
        else 
            numSpiralPartsWithSize <- (numSpiralPartsWithSize - 1)
    [ 
        let height = matrix.GetLength 0
        for row in 0..height - 1  do
        yield matrix.[row,*] |> List.ofArray
    ]